### PR TITLE
more pullups for 7.7

### PR DESCRIPTION
**Pull request by jaltman**
klist display all known flags

use memset_s instead of memset where appropriate

kadmind disable sync during database load

### CODE REVIEW

**Comment by jaltman** [comment]
@nicowilliams please review conflict resolution in cherry-pick of 7d5f8bb051ca84592d1196bf5d5522da5a50f9d6

File: kadmin/load.c [hunk 1]
```diff
@@ -607,6 +608,9 @@ doit(const char *filename, int mergep)
 	}
     }
     free(line);
+    ret2 = db->hdb_set_sync(context, db, 1);
+    if (ret2)
+        krb5_err(context, 1, ret, "failed to sync the HDB");
```
Comment (lhoward) [inline kadmin/load.c]: Can you coalesce these as done in master, i.e.

```
    ret2 = db->hdb_set_sync(context, db, 1);
    if (ret2) {
        krb5_err(context, 1, ret2, "failed to sync the HDB");
        ret = ret2;
    }
```

**Comment by quanah** [comment]
I see the following warnings with memset_s on Solaris 11.4:

```
/export/home/build/git/sold/sold-master/heimdal/lib/hcrypto/md2.c: In function ‘calc’:
/export/home/build/git/sold/sold-master/heimdal/lib/hcrypto/md2.c:93:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(x, sizeof(x), 0, sizeof(x));
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/lib/hcrypto/md2.c:93:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/hcrypto/rand-fortuna.c: In function ‘md_result’:
/export/home/build/git/sold/sold-master/heimdal/lib/hcrypto/rand-fortuna.c:162:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(&tmp, sizeof(tmp), 0, sizeof(tmp));
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/lib/hcrypto/rand-fortuna.c:162:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/hx509/ks_file.c: In function ‘try_decrypt’:
/export/home/build/git/sold/sold-master/heimdal/lib/hx509/ks_file.c:129:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(clear.data, clear.length, 0, clear.length);
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/lib/hx509/ks_file.c:129:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/crypto.c: In function ‘encrypt_internal_enc_then_cksum’:
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/crypto.c:941:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(p, total_sz, 0, total_sz);
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/crypto.c:941:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/crypto-arcfour.c: In function ‘ARCFOUR_subencrypt’:
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/crypto-arcfour.c:183:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(k1_c_data, sizeof(k1_c_data), 0, sizeof(k1_c_data));
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/crypto-arcfour.c:183:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/crypto-des-common.c: In function ‘_krb5_des_checksum’:
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/crypto-des-common.c:80:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(&ivec, sizeof(ivec), 0, sizeof(ivec));
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/crypto-des-common.c:80:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/crypto-pk.c: In function ‘_krb5_pk_octetstring2key’:
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/crypto-pk.c:95:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(shaoutput, sizeof(shaoutput), 0, sizeof(shaoutput));
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/crypto-pk.c:95:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/deprecated.c: In function ‘krb5_password_key_proc’:
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/deprecated.c:210:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(buf, sizeof(buf), 0, sizeof(buf));
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/deprecated.c:210:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/init_creds_pw.c: In function ‘free_init_creds_ctx’:
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/init_creds_pw.c:167:2: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
  memset_s(ctx->password, len, 0, len);
  ^~~~~~~~
  memset
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/init_creds_pw.c:167:2: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/kcm.c: In function ‘kcm_free’:
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/kcm.c:227:2: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
  memset_s(k, sizeof(*k), 0, sizeof(*k));
  ^~~~~~~~
  memset
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/kcm.c:227:2: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```
```
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/salt-des.c: In function ‘krb5_DES_AFS3_Transarc_string_to_key’:
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/salt-des.c:112:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(password, sizeof(password), 0, sizeof(password));
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/salt-des.c:112:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/salt-des3.c: In function ‘DES3_string_to_key’:
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/salt-des3.c:64:6: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
      memset_s(str, len, 0, len);
      ^~~~~~~~
      memset
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/salt-des3.c:64:6: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/crypto-pk.c: In function ‘_krb5_pk_octetstring2key’:
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/crypto-pk.c:95:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(shaoutput, sizeof(shaoutput), 0, sizeof(shaoutput));
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/crypto-pk.c:95:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/salt-des.c: In function ‘krb5_DES_AFS3_Transarc_string_to_key’:
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/salt-des.c:112:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(password, sizeof(password), 0, sizeof(password));
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/lib/krb5/salt-des.c:112:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/ntlm/ntlm.c: In function ‘splitandenc’:
/export/home/build/git/sold/sold-master/heimdal/lib/ntlm/ntlm.c:1195:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(key, sizeof(key), 0, sizeof(key));
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/lib/ntlm/ntlm.c:1195:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/gssapi/krb5/arcfour.c: In function ‘_gssapi_get_mic_arcfour’:
/export/home/build/git/sold/sold-master/heimdal/lib/gssapi/krb5/arcfour.c:307:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(k6_data, sizeof(k6_data), 0, sizeof(k6_data));
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/lib/gssapi/krb5/arcfour.c:307:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/gssapi/krb5/get_mic.c: In function ‘mic_des’:
/export/home/build/git/sold/sold-master/heimdal/lib/gssapi/krb5/get_mic.c:124:3: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
   memset_s(deskey, sizeof(deskey), 0, sizeof(deskey));
   ^~~~~~~~
   memset
/export/home/build/git/sold/sold-master/heimdal/lib/gssapi/krb5/get_mic.c:124:3: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/gssapi/krb5/verify_mic.c: In function ‘verify_mic_des’:
/export/home/build/git/sold/sold-master/heimdal/lib/gssapi/krb5/verify_mic.c:92:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(deskey, sizeof(deskey), 0, sizeof(deskey));
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/lib/gssapi/krb5/verify_mic.c:92:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/gssapi/ntlm/init_sec_context.c: In function ‘from_file’:
/export/home/build/git/sold/sold-master/heimdal/lib/gssapi/ntlm/init_sec_context.c:75:2: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
  memset_s(buf, sizeof(buf), 0, sizeof(buf));
  ^~~~~~~~
  memset
/export/home/build/git/sold/sold-master/heimdal/lib/gssapi/ntlm/init_sec_context.c:75:2: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/lib/hdb/mkey.c: In function ‘read_master_encryptionkey’:
/export/home/build/git/sold/sold-master/heimdal/lib/hdb/mkey.c:229:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(buf, sizeof(buf), 0, sizeof(buf));
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/lib/hdb/mkey.c:229:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/kuser/kinit.c: In function ‘get_new_tickets’:
/export/home/build/git/sold/sold-master/heimdal/kuser/kinit.c:718:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(passwd, sizeof(passwd), 0, sizeof(passwd));
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/kuser/kinit.c:718:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/kadmin/cpw.c: In function ‘set_random_password’:
/export/home/build/git/sold/sold-master/heimdal/kadmin/cpw.c:79:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(pw, sizeof(pw), 0, sizeof(pw));
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/kadmin/cpw.c:79:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/kadmin/random_password.c: In function ‘random_password’:
/export/home/build/git/sold/sold-master/heimdal/kadmin/random_password.c:72:5: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
     memset_s(pass, len, 0, len);
     ^~~~~~~~
     memset
/export/home/build/git/sold/sold-master/heimdal/kadmin/random_password.c:72:5: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/kadmin/server.c: In function ‘kadmind_dispatch’:
/export/home/build/git/sold/sold-master/heimdal/kadmin/server.c:576:2: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
  memset_s(password, len, 0, len);
  ^~~~~~~~
  memset
/export/home/build/git/sold/sold-master/heimdal/kadmin/server.c:576:2: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

```
/export/home/build/git/sold/sold-master/heimdal/appl/su/su.c:306:2: warning: implicit declaration of function ‘memset_s’; did you mean ‘memset’? [-Wimplicit-function-declaration]
  memset_s(pw_buf, sizeof(pw_buf), 0, sizeof(pw_buf));
  ^~~~~~~~
  memset
/export/home/build/git/sold/sold-master/heimdal/appl/su/su.c:306:2: warning: nested extern declaration of ‘memset_s’ [-Wnested-externs]
```

**Comment by quanah** [comment]
After adding:

-D_STDC_C11_BCI to the CPPFLAGS when I compiled, the warnings went away.

**Comment by nicowilliams** [comment]
I'm OK with adding that define, `-D_STDC_C11_BCI` for this release, but since I can't find anything about it, for 8.0 I'd rather have an autoconf macro that looks for `memset_s()` being available in the C library w/ or w/o a prototype defined in the headers, and if not, then enable a prototype to be provided by `lib/roken`.

Have you tried just using `-stdc=11`?

File: kadmin/load.c (hunk 1 above)
Comment (jaltman) [reply kadmin/load.c]: They shouldn't have been coalesced.  The hdb_set_sync() call needed to be inserted after the `if (ret2) ret = ret2;` statement.

**Comment by jaltman** [comment]
Proposed patch to address Solaris memset_s warnings.
`diff --git a/cf/sunos.m4 b/cf/sunos.m4
index 57a244e90..938fc3a5e 100644
--- a/cf/sunos.m4
+++ b/cf/sunos.m4
@@ -11,6 +11,8 @@ case "$host" in
 *-*-solaris2.1[[1-9]])
        AC_DEFINE([__EXTENSIONS__], [1],
                  [Enable general extensions on Solaris.])
+       AC_DEFINE([_STDC_C11_BCI], [1],
+               [Enable C11 prototypes for memset_s and friends])
        sunos=511
        ;;
 *-*-solaris2.[[89]] | *-*-solaris2.10)
`

**Comment by quanah** [comment]
> Have you tried just using `-stdc=11`?

I'm using gcc on Solaris 11 (It's the default compiler with Solaris 11.4 at least, I think on 11.3 as well), where the option is `--std=c11`

If I set that option, the build fails with:

```
/export/home/build/git/sold/sold-master/heimdal/appl/dbutils/bsearch.c:133:13: error: ‘optind’ undeclared (first use in this function); did you mean ‘optidx’?
     argc -= optind;
             ^~~~~~
             optidx
/export/home/build/git/sold/sold-master/heimdal/appl/dbutils/bsearch.c:133:13: note: each undeclared identifier is reported only once for each function it appears in
```

Which is somewhat odd, since the configure script found it:

```
config.log:configure:24448: checking whether optind is declared
config.log:ac_cv_have_decl_optind=yes
```

and sets

```
#define HAVE_DECL_OPTIND 1
```

Previously I was using `--std=gnu99` and did not encounter this issue.

In looking at the header files in /usr/include, I see:

```
#if defined(__EXTENSIONS__) || \
        !defined(_STRICT_STDC) || defined(_XOPEN_SOURCE)
/* Marked LEGACY in SUSv2 and removed in SUSv3 */
#if !defined(_XPG6) || defined(__EXTENSIONS__)
extern char     *cuserid(char *);
extern int      getopt(int, char *const *, const char *);
extern char     *optarg;
extern int      optind, opterr, optopt, optreset;
extern int      getw(FILE *);
extern int      putw(int, FILE *);
#endif /* !defined(_XPG6) || defined(__EXTENSIONS__) */
#endif
```

So it looks like enabling C11 disables necessary bits for use with compilation.

**Comment by jaltman** [comment]
f3b74d8d72436e928755471a704384e33a33ef72 defines _STDC_C11_BCI on Solaris

**Pull request merged**